File: 77-combinations/combinations.py
class Solution:
    def combine(self, n: int, k: int) -> List[List[int]]:
        def myRec(index, arr):
            if len(arr) == k:
                res.append(arr[:])
                return
            if index > n or len(arr) > k:
                return
            arr.append(index)
            myRec(index + 1, arr)
            arr.pop()
            myRec(index + 1, arr)
            return
        res = []
        myRec(1,[])
        return res

File: 77-combinations/combinations.java
class Solution {
    public List<List<Integer>> result;
    public int N;
    public int K;
    public List<List<Integer>> combine(int n, int k) {
        result = new ArrayList<>();
        N = n;
        K = k;
        fun(new ArrayList<>(),1);
        return result;
    }
    public void fun(List<Integer> curr, int index){ 
        if(curr.size() == K){
            result.add(new ArrayList<>(curr));
            return;
        }
        if(index > N || curr.size() > K){
            return;
        }
        curr.add(index);
        fun(curr,index+1);
        curr.remove(curr.size()-1);
        fun(curr,index + 1);
    }
}